File: templates/wasm-node-python/examples/string_nodes.py
"""
String Nodes — Text manipulation utilities

Demonstrates string processing nodes including case conversion,
trimming, length analysis, search, replace, concat, and reverse.
"""

from sdk import (
    Context,
    ExecutionResult,
    NodeDefinition,
    PinDefinition,
    PinType,
)


def get_definitions() -> list[NodeDefinition]:
    nodes: list[NodeDefinition] = []

    for name, friendly, desc in [
        ("string_uppercase_py", "To Uppercase", "Converts text to uppercase"),
        ("string_lowercase_py", "To Lowercase", "Converts text to lowercase"),
        ("string_trim_py", "Trim", "Removes leading and trailing whitespace"),
        ("string_reverse_py", "Reverse", "Reverses the characters in a string"),
    ]:
        nd = NodeDefinition(name, friendly, desc, "String/Transform")
        nd.add_pin(PinDefinition.input_exec("exec"))
        nd.add_pin(PinDefinition.input_pin("text", PinType.STRING, default=""))
        nd.add_pin(PinDefinition.output_exec("exec_out"))
        nd.add_pin(PinDefinition.output_pin("result", PinType.STRING))
        nodes.append(nd)

    # Length
    length = NodeDefinition("string_length_py", "String Length", "Returns the length of a string", "String/Analysis")
    length.add_pin(PinDefinition.input_exec("exec"))
    length.add_pin(PinDefinition.input_pin("text", PinType.STRING, default=""))
    length.add_pin(PinDefinition.output_exec("exec_out"))
    length.add_pin(PinDefinition.output_pin("length", PinType.I64))
    length.add_pin(PinDefinition.output_pin("is_empty", PinType.BOOL))
    nodes.append(length)

    # Contains
    contains = NodeDefinition("string_contains_py", "Contains", "Checks if text contains a substring", "String/Analysis")
    contains.add_pin(PinDefinition.input_exec("exec"))
    contains.add_pin(PinDefinition.input_pin("text", PinType.STRING, default=""))
    contains.add_pin(PinDefinition.input_pin("search", PinType.STRING, default=""))
    contains.add_pin(PinDefinition.output_exec("exec_out"))
    contains.add_pin(PinDefinition.output_pin("result", PinType.BOOL))
    nodes.append(contains)

    # Replace
    replace = NodeDefinition("string_replace_py", "Replace", "Replaces occurrences of a pattern", "String/Transform")
    replace.add_pin(PinDefinition.input_exec("exec"))
    replace.add_pin(PinDefinition.input_pin("text", PinType.STRING, default=""))
    replace.add_pin(PinDefinition.input_pin("find", PinType.STRING, default=""))
    replace.add_pin(PinDefinition.input_pin("replace_with", PinType.STRING, default=""))
    replace.add_pin(PinDefinition.output_exec("exec_out"))
    replace.add_pin(PinDefinition.output_pin("result", PinType.STRING))
    replace.add_pin(PinDefinition.output_pin("count", PinType.I64))
    nodes.append(replace)

    # Concat
    concat = NodeDefinition("string_concat_py", "Concatenate", "Joins two strings together", "String/Transform")
    concat.add_pin(PinDefinition.input_exec("exec"))
    concat.add_pin(PinDefinition.input_pin("a", PinType.STRING, default=""))
    concat.add_pin(PinDefinition.input_pin("b", PinType.STRING, default=""))
    concat.add_pin(PinDefinition.input_pin("separator", PinType.STRING, default=""))
    concat.add_pin(PinDefinition.output_exec("exec_out"))
    concat.add_pin(PinDefinition.output_pin("result", PinType.STRING))
    nodes.append(concat)

    return nodes


def run_uppercase(ctx: Context) -> ExecutionResult:
    text = ctx.get_string("text", "")
    ctx.set_output("result", text.upper())
    return ctx.success()


def run_lowercase(ctx: Context) -> ExecutionResult:
    text = ctx.get_string("text", "")
    ctx.set_output("result", text.lower())
    return ctx.success()


def run_trim(ctx: Context) -> ExecutionResult:
    text = ctx.get_string("text", "")
    ctx.set_output("result", text.strip())
    return ctx.success()


def run_reverse(ctx: Context) -> ExecutionResult:
    text = ctx.get_string("text", "")
    ctx.set_output("result", text[::-1])
    return ctx.success()


def run_length(ctx: Context) -> ExecutionResult:
    text = ctx.get_string("text", "")
    ctx.set_output("length", len(text))
    ctx.set_output("is_empty", len(text) == 0)
    return ctx.success()


def run_contains(ctx: Context) -> ExecutionResult:
    text = ctx.get_string("text", "")
    search = ctx.get_string("search", "")
    ctx.set_output("result", search in text)
    return ctx.success()


def run_replace(ctx: Context) -> ExecutionResult:
    text = ctx.get_string("text", "")
    find = ctx.get_string("find", "")
    replace_with = ctx.get_string("replace_with", "")
    count = text.count(find) if find else 0
    result = text.replace(find, replace_with) if find else text
    ctx.set_output("result", result)
    ctx.set_output("count", count)
    return ctx.success()


def run_concat(ctx: Context) -> ExecutionResult:
    a = ctx.get_string("a", "")
    b = ctx.get_string("b", "")
    separator = ctx.get_string("separator", "")
    result = f"{a}{separator}{b}" if separator else f"{a}{b}"
    ctx.set_output("result", result)
    return ctx.success()


DISPATCH = {
    "string_uppercase_py": run_uppercase,
    "string_lowercase_py": run_lowercase,
    "string_trim_py": run_trim,
    "string_reverse_py": run_reverse,
    "string_length_py": run_length,
    "string_contains_py": run_contains,
    "string_replace_py": run_replace,
    "string_concat_py": run_concat,
}


def run(node_name: str, ctx: Context) -> ExecutionResult:
    handler = DISPATCH.get(node_name)
    if handler is None:
        return ctx.fail(f"Unknown node: {node_name}")
    return handler(ctx)

File: templates/wasm-node-assemblyscript/examples/string_nodes.ts
import {
	DataType,
	ExecutionResult,
	NodeDefinition,
	PinDefinition,
	Context,
	FlowNode,
	NodePackage,
} from "@flow-like/wasm-sdk/assembly/index";

export { alloc, dealloc, get_abi_version } from "@flow-like/wasm-sdk/assembly/index";

export class UppercaseNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "string_uppercase_as";
		def.friendly_name = "To Uppercase (AS)";
		def.description = "Converts text to uppercase";
		def.category = "String/Transform";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("text", "Text", "Text to convert", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "Uppercase text", DataType.String));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setString("result", ctx.getString("text").toUpperCase());
		return ctx.success();
	}
}

export class LowercaseNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "string_lowercase_as";
		def.friendly_name = "To Lowercase (AS)";
		def.description = "Converts text to lowercase";
		def.category = "String/Transform";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("text", "Text", "Text to convert", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "Lowercase text", DataType.String));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setString("result", ctx.getString("text").toLowerCase());
		return ctx.success();
	}
}

export class TrimNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "string_trim_as";
		def.friendly_name = "Trim (AS)";
		def.description = "Removes leading and trailing whitespace";
		def.category = "String/Transform";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("text", "Text", "Text to trim", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "Trimmed text", DataType.String));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setString("result", ctx.getString("text").trim());
		return ctx.success();
	}
}

export class LengthNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "string_length_as";
		def.friendly_name = "String Length (AS)";
		def.description = "Returns the length of a string";
		def.category = "String/Analysis";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("text", "Text", "Text to measure", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("length", "Length", "Number of characters", DataType.I64));
		def.addPin(PinDefinition.output("is_empty", "Is Empty", "True if string is empty", DataType.Bool));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		const text = ctx.getString("text");
		ctx.setI64("length", text.length);
		ctx.setBool("is_empty", text.length == 0);
		return ctx.success();
	}
}

export class ContainsNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "string_contains_as";
		def.friendly_name = "Contains (AS)";
		def.description = "Checks if text contains a substring";
		def.category = "String/Analysis";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("text", "Text", "Text to search in", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.input("search", "Search", "Substring to find", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Found", "True if substring found", DataType.Bool));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setBool("result", ctx.getString("text").includes(ctx.getString("search")));
		return ctx.success();
	}
}

export class ReplaceNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "string_replace_as";
		def.friendly_name = "Replace (AS)";
		def.description = "Replaces occurrences of a pattern";
		def.category = "String/Transform";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("text", "Text", "Original text", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.input("find", "Find", "Pattern to find", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.input("replace_with", "Replace With", "Replacement text", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "Modified text", DataType.String));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setString("result", ctx.getString("text").replaceAll(ctx.getString("find"), ctx.getString("replace_with")));
		return ctx.success();
	}
}

export class ConcatNode extends FlowNode {
	define(): NodeDefinition {
		const def = new NodeDefinition();
		def.name = "string_concat_as";
		def.friendly_name = "Concatenate (AS)";
		def.description = "Joins two strings together";
		def.category = "String/Transform";

		def.addPin(PinDefinition.input("exec", "Execute", "Trigger", DataType.Exec));
		def.addPin(PinDefinition.input("a", "A", "First string", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.input("b", "B", "Second string", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.input("separator", "Separator", "Text between strings", DataType.String).withDefaultString(""));
		def.addPin(PinDefinition.output("exec_out", "Done", "Complete", DataType.Exec));
		def.addPin(PinDefinition.output("result", "Result", "Combined string", DataType.String));
		return def;
	}

	execute(ctx: Context): ExecutionResult {
		ctx.setString("result", ctx.getString("a") + ctx.getString("separator") + ctx.getString("b"));
		return ctx.success();
	}
}

const pkg = new NodePackage();
pkg.register(new UppercaseNode());
pkg.register(new LowercaseNode());
pkg.register(new TrimNode());
pkg.register(new LengthNode());
pkg.register(new ContainsNode());
pkg.register(new ReplaceNode());
pkg.register(new ConcatNode());

export function get_nodes(): i64 {
	return pkg.getNodes();
}

export function run(ptr: i32, len: i32): i64 {
	return pkg.run(ptr, len);
}
